File: backend/app/models/documentation.py
import json
from dataclasses import dataclass, field

from pydantic import BaseModel

from app.models.knowledge import JobStatus


class DiscoveredPage(BaseModel):
    url: str
    title: str


class DocumentationDiscoverRequest(BaseModel):
    url: str
    user_id: str
    use_cookies: bool = True


class DocumentationDiscoverResponse(BaseModel):
    entry_url: str
    scope_path: str
    site_name: str
    pages: list[DiscoveredPage]
    total_count: int
    truncated: bool
    original_count: int | None = None
    has_cookies: bool


class DocumentationScrapeRequest(BaseModel):
    user_id: str
    entry_url: str
    site_name: str
    scope_path: str
    pages: list[DiscoveredPage]
    use_cookies: bool = True


class DocumentationScrapeResponse(BaseModel):
    job_id: str
    collection_id: str
    message: str


class DocumentationRetryResponse(BaseModel):
    job_id: str
    collection_id: str
    retry_count: int
    message: str


class DocumentationPageModel(BaseModel):
    id: str
    page_url: str
    title: str | None
    status: str
    is_truncated: bool
    display_order: int


class DocumentationPagesResponse(BaseModel):
    collection_id: str
    pages: list[DocumentationPageModel]


class DocumentationDeleteResponse(BaseModel):
    message: str
    pages_deleted: int
    vectors_deleted: bool


@dataclass
class DocScrapeJob:
    """Job tracker for documentation scraping SSE dispatch."""

    id: str
    status: JobStatus = JobStatus.PENDING
    total_pages: int = 0
    processed_pages: int = 0
    failed_pages: list[str] = field(default_factory=list)
    succeeded_pages: list[str] = field(default_factory=list)
    message: str = ""

    @property
    def progress(self) -> int:
        if self.total_pages == 0:
            return 0
        return int((self.processed_pages / self.total_pages) * 100)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "status": self.status.value,
            "progress": self.progress,
            "total_pages": self.total_pages,
            "processed_pages": self.processed_pages,
            "failed_pages": self.failed_pages,
            "succeeded_pages": self.succeeded_pages,
            "message": self.message,
        }

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

File: next-frontend/lib/types/documentation.ts
export interface DiscoveredPage {
  url: string;
  title: string;
}

export interface DocumentationCollection {
  id: string;
  user_id: string;
  entry_url: string;
  site_name: string | null;
  scope_path: string;
  total_pages: number;
  successful_pages: number;
  status: 'discovering' | 'pending_confirmation' | 'scraping' | 'completed' | 'partial' | 'failed';
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentationPage {
  id: string;
  collection_id: string;
  user_id: string;
  page_url: string;
  title: string | null;
  content_markdown: string | null;
  status: 'pending' | 'scraping' | 'completed' | 'failed';
  error_message: string | null;
  is_truncated: boolean;
  display_order: number;
  created_at: string;
}

export interface DiscoveryResponse {
  entry_url: string;
  scope_path: string;
  site_name: string;
  pages: DiscoveredPage[];
  total_count: number;
  truncated: boolean;
  original_count?: number;
  has_cookies: boolean;
}

export interface ScrapeResponse {
  job_id: string;
  collection_id: string;
  message: string;
}

export interface RetryResponse {
  job_id: string;
  collection_id: string;
  retry_count: number;
  message: string;
}

export interface DocJobStatusUpdate {
  id: string;
  status: string;
  progress: number;
  total_pages: number;
  processed_pages: number;
  failed_pages: string[];
  succeeded_pages: string[];
  message: string;
}
